### DISCUSSION TITLE
Help with TypeScript AST traversal

### DISCUSSION THREAD

**Discussion by F0rsaken** [vuejs/language-tools#1932] [General]
Hi, my question is loosly Volar related and I will delete this discussion if you think it makes a mess or smth.

I'm building VSCode extension for small framework my team uses, and I need help with detecting and suggesting components, as well as Go To Definition feature. I don't know if it's a good place to ask, but I thought I will try either way.

## My problem
**_Scanning and detecting all components that are registered in app_**

After I deal with that problem, I can do anything. I know Volar uses this syntax to declare global components:

```ts
declare module '@vue/runtime-core' {
	export interface GlobalComponents {
		MyComponent: typeof import('@/components/MyComponent.vue')
	}
}
```
I love this and I would like to know how Volar is doing it, how it reads this interface and detects all components, or at least where in the source code I could find it.

### More framework specific details

In my framework, I register components like this:

```ts
const app = new App({
	components: [ComponentA, ComponentB]
})
```

But often there is a `components` array, that gather them all. I'm not sure if reading a type from it will do and will help me find definitions of those classes. I tried leveraging TS Compiler API, but I had hard time with it.

I would love any help in this matter, or any learning resource that would bring me a little closer to my solution.

**Comment by johnsoncodehk** [comment]
We have try to get component types from `GlobalComponents` interface in virtual code, you can run `Volar (Debug): Show Virtual Files` command to see.
<img width="1361" alt="螢幕截圖 2022-11-28 17 45 20" src="https://user-images.githubusercontent.com/16279759/204245970-c57a5979-fb61-4dd0-88c9-22b927d8159d.png">
